File: DevExpress VCL/ExpressPivotGrid 2/Demos/CBuilder/Common/DemoBasicMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "DemoBasicMain.h"
#include "cxExportPivotGridLink.hpp"
#include "AboutDemoForm.h"
#include "DemoUtils.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxLookAndFeels"
#pragma resource "*.dfm"

//---------------------------------------------------------------------------
__fastcall TfrmDemoBaisicMain::TfrmDemoBaisicMain(TComponent* Owner)
  : TForm(Owner)
{
  Application->Title = Caption;
  SyncMenuWithTotalVisibility();
  SyncMenuWithElementsVisibility();
  SyncMenuWithOptionsSelection();
}

TcxCustomPivotGrid* __fastcall TfrmDemoBaisicMain::PivotGrid()
{
  return NULL;
}

void __fastcall TfrmDemoBaisicMain::miExportToClick(TObject *Sender)
{
  if (SaveDialog->Execute()) {
	switch (((TMenuItem*)Sender)->Tag){
	  case 1:
		cxExportPivotGridToExcel(SaveDialog->FileName, PivotGrid(), true, true, "xls"); break;
	  case 2:
		cxExportPivotGridToText(SaveDialog->FileName, PivotGrid(), true, "", "", "", "txt"); break;
	  case 3:
		cxExportPivotGridToHTML(SaveDialog->FileName, PivotGrid(), true, "html"); break;
	  case 4:
		cxExportPivotGridToXML(SaveDialog->FileName, PivotGrid(), true, "xml"); break;
	}
  }
}

void __fastcall TfrmDemoBaisicMain::miTotalsLocationClick(TObject *Sender)
{
	switch(((TMenuItem*)Sender)->Tag)
	{
		case 0, 1:
			PivotGrid()->OptionsView->ColumnTotalsLocation = TcxPivotGridTotalsLocation(((TMenuItem*)Sender)->Tag);
		case 2, 3:
			PivotGrid()->OptionsView->RowTotalsLocation = TcxPivotGridTotalsLocation(((TMenuItem*)Sender)->Tag - 2);
	}
	miColumnTotalsPositionFar->Checked = PivotGrid()->OptionsView->ColumnTotalsLocation == tlFar;
	miColumnTotalsPositionNear->Checked = PivotGrid()->OptionsView->ColumnTotalsLocation == tlNear;
	miRowTotalsPositionFar->Checked = PivotGrid()->OptionsView->RowTotalsLocation == tlFar;
	miRowTotalsPositionNear->Checked = PivotGrid()->OptionsView->RowTotalsLocation == tlNear;
}

void __fastcall TfrmDemoBaisicMain::miExitClick(TObject *Sender)
{
  Application->Terminate();
}

void __fastcall TfrmDemoBaisicMain::FormCreate(TObject *Sender)
{
  SetDefaultLookAndFeel();
  AddLookAndFeelMenu();
}

void __fastcall TfrmDemoBaisicMain::miTotalsVisibilityClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  SyncTotalVisibilityWithMenu();
}

void __fastcall TfrmDemoBaisicMain::miElementsVisibilityClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  SyncElementsVisibilityWithMenu();
}
void __fastcall TfrmDemoBaisicMain::miAboutClick(TObject *Sender)
{
  ShowAboutDemoForm();
}

void __fastcall TfrmDemoBaisicMain::SyncElementsVisibilityWithMenu()
{
  PivotGrid()->OptionsView->ColumnFields = miShowColumnFields->Checked;
  PivotGrid()->OptionsView->DataFields = miShowDataFields->Checked;
  PivotGrid()->OptionsView->FilterFields = miShowFilterFields->Checked;
  PivotGrid()->OptionsView->FilterSeparator = miShowFilterSeparator->Checked;
  PivotGrid()->OptionsView->RowFields = miShowRowFields->Checked;
}

void __fastcall TfrmDemoBaisicMain::SyncTotalVisibilityWithMenu()
{
  PivotGrid()->OptionsView->ColumnTotals = miShowColumnTotals->Checked;
  PivotGrid()->OptionsView->RowTotals = miShowRowTotals->Checked;
  PivotGrid()->OptionsView->ColumnGrandTotals = miShowColumnGrandTotals->Checked;
  PivotGrid()->OptionsView->RowGrandTotals = miShowRowGrandTotals->Checked;
  PivotGrid()->OptionsView->TotalsForSingleValues = miShowTotalsForSingleValues->Checked;
  PivotGrid()->OptionsView->GrandTotalsForSingleValues = miShowGrandTotalsForSingleValues->Checked;
}

void __fastcall TfrmDemoBaisicMain::SyncMenuWithElementsVisibility()
{
  miShowColumnFields->Checked = PivotGrid()->OptionsView->ColumnFields;
  miShowDataFields->Checked = PivotGrid()->OptionsView->DataFields;
  miShowFilterFields->Checked = PivotGrid()->OptionsView->FilterFields;
  miShowFilterSeparator->Checked = PivotGrid()->OptionsView->FilterSeparator;
  miShowRowFields->Checked = PivotGrid()->OptionsView->RowFields;
}

void __fastcall TfrmDemoBaisicMain::SyncMenuWithOptionsSelection()
{
  miMultiSelect->Checked = PivotGrid()->OptionsSelection->MultiSelect;
  miCrossCells->Checked = PivotGrid()->OptionsSelection->IncludeCells.Contains(osiCrossCells);
  miGrandTotalsCells->Checked = PivotGrid()->OptionsSelection->IncludeCells.Contains(osiGrandTotalCells);
  miTotalsCells->Checked = PivotGrid()->OptionsSelection->IncludeCells.Contains(osiTotalCells);
  miHideFocusRect->Checked = PivotGrid()->OptionsSelection->HideFocusRect;
  miHideSelection->Checked = PivotGrid()->OptionsSelection->HideSelection;
}

void __fastcall TfrmDemoBaisicMain::SyncMenuWithTotalVisibility()
{
  miShowColumnTotals->Checked = PivotGrid()->OptionsView->ColumnTotals;
  miShowRowTotals->Checked = PivotGrid()->OptionsView->RowTotals;
  miShowColumnGrandTotals->Checked = PivotGrid()->OptionsView->ColumnGrandTotals;
  miShowRowGrandTotals->Checked = PivotGrid()->OptionsView->RowGrandTotals;
  miShowTotalsForSingleValues->Checked = PivotGrid()->OptionsView->TotalsForSingleValues;
  miShowGrandTotalsForSingleValues->Checked = PivotGrid()->OptionsView->GrandTotalsForSingleValues;
}

void __fastcall TfrmDemoBaisicMain::miMultiSelectClick(TObject *Sender)
{
  PivotGrid()->OptionsSelection->MultiSelect = !PivotGrid()->OptionsSelection->MultiSelect;
  ((TMenuItem*)Sender)->Checked = PivotGrid()->OptionsSelection->MultiSelect;
}

void __fastcall TfrmDemoBaisicMain::miHideFocusRectClick(TObject *Sender)
{
  PivotGrid()->OptionsSelection->HideFocusRect = !PivotGrid()->OptionsSelection->HideFocusRect;
  ((TMenuItem*)Sender)->Checked = PivotGrid()->OptionsSelection->HideFocusRect;
}

void __fastcall TfrmDemoBaisicMain::miHideSelectionClick(TObject *Sender)
{
  PivotGrid()->OptionsSelection->HideSelection = !PivotGrid()->OptionsSelection->HideSelection;
  ((TMenuItem*)Sender)->Checked = PivotGrid()->OptionsSelection->HideSelection;
}

void __fastcall TfrmDemoBaisicMain::IncludeCellsClick(TObject *Sender)
{
  TcxPivotGridOptionsSelectionIncludes AIncludeCells;

  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  AIncludeCells.Clear();
  if (miCrossCells->Checked)
	AIncludeCells << osiCrossCells;
  if (miGrandTotalsCells->Checked)
	AIncludeCells << osiGrandTotalCells;
  if (miTotalsCells->Checked)
	AIncludeCells << osiTotalCells;
  PivotGrid()->OptionsSelection->IncludeCells = AIncludeCells;
}

void __fastcall TfrmDemoBaisicMain::AddLookAndFeelMenu()
{
  mmMain->Items->Insert(mmMain->Items->IndexOf(miAbout),
	CreateLookAndFeelMenuItems(mmMain->Items, lfController));
}

TcxLookAndFeelKind __fastcall TfrmDemoBaisicMain::GetDefaultLookAndFeelKind()
{
  return(lfOffice11);
}

bool __fastcall TfrmDemoBaisicMain::IsNativeDefaultStyle()
{
  return(false);
}

void __fastcall TfrmDemoBaisicMain::SetDefaultLookAndFeel()
{
  lfController->NativeStyle = IsNativeDefaultStyle();
  lfController->Kind = GetDefaultLookAndFeelKind();
}
//---------------------------------------------------------------------------

File: DevExpress VCL/ExpressPivotGrid 2/Demos/Delphi/Common/DemoBasicMain.pas
unit DemoBasicMain;

{$I ..\cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxLookAndFeels, StdCtrls, Menus, cxCustomPivotGrid, cxExportPivotGridLink;

type
  TfrmDemoBaisicMain = class(TForm)
    mmMain: TMainMenu;
    miFile: TMenuItem;
    miExport: TMenuItem;
    miExportToExcel: TMenuItem;
    miExportToText: TMenuItem;
    miExportToHtml: TMenuItem;
    miExportToXml: TMenuItem;
    Separator1: TMenuItem;
    miExit: TMenuItem;
    miOptions: TMenuItem;
    miAbout: TMenuItem;
    SaveDialog: TSaveDialog;
    lbDescrip: TLabel;
    lfController: TcxLookAndFeelController;
    miTotalsPosition: TMenuItem;
    miShowTotalsForSingleValues: TMenuItem;
    miShowRowTotals: TMenuItem;
    miShowColumnTotals: TMenuItem;
    miShowColumnGrandTotals: TMenuItem;
    miShowRowGrandTotals: TMenuItem;
    miTotalsVisibility: TMenuItem;
    miShowGrandTotalsForSingleValues: TMenuItem;
    N2: TMenuItem;
    miElementsVisibility: TMenuItem;
    miShowFilterFields: TMenuItem;
    miShowColumnFields: TMenuItem;
    miShowDataFields: TMenuItem;
    miShowRowFields: TMenuItem;
    miShowFilterSeparator: TMenuItem;
    N1: TMenuItem;
    miColumnTotalsPosition: TMenuItem;
    miRowTotalsPosition: TMenuItem;
    miColumnTotalsPositionFar: TMenuItem;
    miColumnTotalsPositionNear: TMenuItem;
    miRowTotalsPositionFar: TMenuItem;
    miRowTotalsPositionNear: TMenuItem;
    miSelection: TMenuItem;
    miMultiSelect: TMenuItem;
    miHideFocusRect: TMenuItem;
    miHideSelection: TMenuItem;
    miIncludeCells: TMenuItem;
    miCrossCells: TMenuItem;
    miGrandTotalsCells: TMenuItem;
    miTotalsCells: TMenuItem;
    procedure miExportToClick(Sender: TObject);
    procedure miTotalsLocationClick(Sender: TObject);
    procedure miExitClick(Sender: TObject);
    procedure miTotalsVisibilityClick(Sender: TObject);
    procedure miElementsVisibilityClick(Sender: TObject);
    procedure miAboutClick(Sender: TObject);
    procedure miMultiSelectClick(Sender: TObject);
    procedure miHideFocusRectClick(Sender: TObject);
    procedure miHideSelectionClick(Sender: TObject);
    procedure IncludeCellsClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { Private declarations }
  protected
    procedure AddLookAndFeelMenu; virtual;
    function GetDefaultLookAndFeelKind: TcxLookAndFeelKind; virtual;
    function GetPivotGrid: TcxCustomPivotGrid; virtual;
    function IsNativeDefaultStyle: Boolean; virtual;
    procedure SetDefaultLookAndFeel; virtual;
    procedure SyncElementsVisibilityWithMenu;
    procedure SyncTotalVisibilityWithMenu;
    procedure SyncMenuWithElementsVisibility;
    procedure SyncMenuWithOptionsSelection;
    procedure SyncMenuWithTotalVisibility;
  public
    constructor Create(AOwner: TComponent); override;

    property PivotGrid: TcxCustomPivotGrid read GetPivotGrid;
  end;

implementation

uses
  AboutDemoForm, DemoUtils;

{$R *.dfm}

constructor TfrmDemoBaisicMain.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
  Application.Title := Caption;
  PivotGrid.ApplyBestFit;
  SyncMenuWithTotalVisibility;
  SyncMenuWithElementsVisibility;
  SyncMenuWithOptionsSelection;
end;

function TfrmDemoBaisicMain.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := nil;
end;

procedure TfrmDemoBaisicMain.SyncElementsVisibilityWithMenu;
begin
  with PivotGrid.OptionsView do
  begin
    ColumnFields := miShowColumnFields.Checked;
    DataFields := miShowDataFields.Checked;
    FilterFields := miShowFilterFields.Checked;
    FilterSeparator := miShowFilterSeparator.Checked;
    RowFields := miShowRowFields.Checked;
  end;
end;

procedure TfrmDemoBaisicMain.SyncTotalVisibilityWithMenu;
begin
  with PivotGrid.OptionsView do
  begin
    ColumnTotals := miShowColumnTotals.Checked;
    RowTotals := miShowRowTotals.Checked;
    ColumnGrandTotals := miShowColumnGrandTotals.Checked;
    RowGrandTotals := miShowRowGrandTotals.Checked;
    TotalsForSingleValues := miShowTotalsForSingleValues.Checked;
    GrandTotalsForSingleValues := miShowGrandTotalsForSingleValues.Checked;
  end;
end;

procedure TfrmDemoBaisicMain.SyncMenuWithElementsVisibility;
begin
  with PivotGrid.OptionsView do
  begin
    miShowColumnFields.Checked := ColumnFields;
    miShowDataFields.Checked := DataFields;
    miShowFilterFields.Checked := FilterFields;
    miShowFilterSeparator.Checked := FilterSeparator;
    miShowRowFields.Checked := RowFields;
  end;
end;

procedure TfrmDemoBaisicMain.SyncMenuWithOptionsSelection;
begin
  with PivotGrid.OptionsSelection do
  begin
    miMultiSelect.Checked := MultiSelect;
    miCrossCells.Checked := osiCrossCells in IncludeCells;
    miGrandTotalsCells.Checked := osiGrandTotalCells in IncludeCells;
    miTotalsCells.Checked := osiTotalCells in IncludeCells;
    miHideFocusRect.Checked := HideFocusRect;
    miHideSelection.Checked := HideSelection;
  end;
end;

procedure TfrmDemoBaisicMain.SyncMenuWithTotalVisibility;
begin
  with PivotGrid.OptionsView do
  begin
    miShowColumnTotals.Checked := ColumnTotals;
    miShowRowTotals.Checked := RowTotals;
    miShowColumnGrandTotals.Checked := ColumnGrandTotals;
    miShowRowGrandTotals.Checked := RowGrandTotals;
    miShowTotalsForSingleValues.Checked := TotalsForSingleValues;
    miShowGrandTotalsForSingleValues.Checked := GrandTotalsForSingleValues;
  end;
end;

procedure TfrmDemoBaisicMain.miExportToClick(Sender: TObject);
begin
  if SaveDialog.Execute then
  begin
    case TMenuItem(Sender).Tag of
      1:
        cxExportPivotGridToExcel(SaveDialog.FileName, PivotGrid);
      2:
        cxExportPivotGridToText(SaveDialog.FileName, PivotGrid);
      3:
        cxExportPivotGridToHTML(SaveDialog.FileName, PivotGrid);
      4:
        cxExportPivotGridToXML(SaveDialog.FileName, PivotGrid);
    end;
  end;
end;

procedure TfrmDemoBaisicMain.miTotalsLocationClick(Sender: TObject);
begin
  case TMenuItem(Sender).Tag of
    0, 1:
      PivotGrid.OptionsView.ColumnTotalsLocation := TcxPivotGridTotalsLocation(TMenuItem(Sender).Tag);
    2, 3:
      PivotGrid.OptionsView.RowTotalsLocation := TcxPivotGridTotalsLocation(TMenuItem(Sender).Tag - 2);
  end;
  miColumnTotalsPositionFar.Checked := PivotGrid.OptionsView.ColumnTotalsLocation = tlFar;
  miColumnTotalsPositionNear.Checked := PivotGrid.OptionsView.ColumnTotalsLocation = tlNear;
  miRowTotalsPositionFar.Checked := PivotGrid.OptionsView.RowTotalsLocation = tlFar;
  miRowTotalsPositionNear.Checked := PivotGrid.OptionsView.RowTotalsLocation = tlNear;
end;

procedure TfrmDemoBaisicMain.miExitClick(Sender: TObject);
begin
  Application.Terminate;
end;

procedure TfrmDemoBaisicMain.miTotalsVisibilityClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  SyncTotalVisibilityWithMenu;
end;

procedure TfrmDemoBaisicMain.miElementsVisibilityClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  SyncElementsVisibilityWithMenu;
end;

procedure TfrmDemoBaisicMain.miAboutClick(Sender: TObject);
begin
  ShowAboutDemoForm;
end;

procedure TfrmDemoBaisicMain.miMultiSelectClick(Sender: TObject);
begin
  PivotGrid.OptionsSelection.MultiSelect := not PivotGrid.OptionsSelection.MultiSelect;
  TMenuItem(Sender).Checked := PivotGrid.OptionsSelection.MultiSelect;
end;

procedure TfrmDemoBaisicMain.miHideFocusRectClick(Sender: TObject);
begin
  PivotGrid.OptionsSelection.HideFocusRect := not PivotGrid.OptionsSelection.HideFocusRect;
  TMenuItem(Sender).Checked := PivotGrid.OptionsSelection.HideFocusRect;
end;

procedure TfrmDemoBaisicMain.miHideSelectionClick(Sender: TObject);
begin
  PivotGrid.OptionsSelection.HideSelection := not PivotGrid.OptionsSelection.HideSelection;
  TMenuItem(Sender).Checked := PivotGrid.OptionsSelection.HideSelection;
end;

procedure TfrmDemoBaisicMain.IncludeCellsClick(Sender: TObject);
var
  AIncludeCells: TcxPivotGridOptionsSelectionIncludes;
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  AIncludeCells := [];
  if miCrossCells.Checked then
    AIncludeCells := AIncludeCells + [osiCrossCells];
  if miGrandTotalsCells.Checked then
    AIncludeCells := AIncludeCells + [osiGrandTotalCells];
  if miTotalsCells.Checked then
    AIncludeCells := AIncludeCells + [osiTotalCells];
  PivotGrid.OptionsSelection.IncludeCells := AIncludeCells;
end;

procedure TfrmDemoBaisicMain.FormCreate(Sender: TObject);
begin
  SetDefaultLookAndFeel;
  AddLookAndFeelMenu;
end;

procedure TfrmDemoBaisicMain.AddLookAndFeelMenu;
begin
  mmMain.Items.Insert(mmMain.Items.IndexOf(miAbout), CreateLookAndFeelMenuItems(mmMain.Items, lfController));
end;

function TfrmDemoBaisicMain.GetDefaultLookAndFeelKind: TcxLookAndFeelKind;
begin
  Result := lfOffice11;
end;

function TfrmDemoBaisicMain.IsNativeDefaultStyle: Boolean;
begin
  Result := False;
end;

procedure TfrmDemoBaisicMain.SetDefaultLookAndFeel;
begin
  lfController.NativeStyle := IsNativeDefaultStyle;
  lfController.Kind := GetDefaultLookAndFeelKind;
end;

end.
